### PR TITLE
Make clients.GatewayClient.get cancellable

### PR DESCRIPTION
**Pull request by joshklop**
## Description

See [here](https://github.com/NethermindEth/juno/pull/542#discussion_r1101229443).

## Changes:

See commits.

## Types of changes

- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)
- Documentation Update
- Code style update (formatting, renaming)
- Refactoring (no functional changes, no api changes)

## Testing

**Requires testing**: Yes

**Did you write tests??**: Existing tests are sufficient

## Documentation

**If this requires a documentation update, did you add one?** No, no

### CODE REVIEW

File: clients/gateway.go [hunk 1]
```diff
@@ -75,15 +78,24 @@ func (c *GatewayClient) buildQueryString(endpoint string, args map[string]string
...
 // get performs a "GET" http request with the given URL and returns the response body
-func (c *GatewayClient) get(queryUrl string) ([]byte, error) {
+func (c *GatewayClient) get(ctx context.Context, queryUrl string) ([]byte, error) {
 	var res *http.Response
 	var err error
+	var body io.Reader
 	wait := c.minWait
 	for i := 0; i <= c.maxRetries; i++ {
-		res, err = http.Get(queryUrl)
+		req, err := http.NewRequestWithContext(ctx, "GET", queryUrl, body)
```
Comment (omerfirmak) [inline clients/gateway.go:87]: we are passing `nil` as the body here, do we need the variable?

File: node/node.go [hunk 2]
```diff
@@ -14,7 +15,7 @@ import (
 )
 
 type StarknetNode interface {
-	Run() error
+	Run(ctx context.Context) error
```
Comment (omerfirmak) [inline node/node.go:18]: We have an explicit `Shutdown()`, so we don't need a `ctx` here. Alternative is getting rid of `Shutdown()`. Same goes for Synchronizer

Comment (omerfirmak) [reply node/node.go:18]: I think getting rid of Shutdown is the better option here btw

Comment (joshklop) [reply node/node.go:18]: That's a great point! Didn't think of that.

**Review by omerfirmak — APPROVED** [review]

**Comment by IronGauntlets** [comment]
I am not an of using context because of the following reasons:

- In most cases, context is used for its quit and done functions. Which are nothing more than closing and reading from a quit channel.
- Due to the context pattern you cannot store it thus you have to pass it to every function in the call stack which may not even call it.
- Context also has a data field which may be used to pass data however all type of safety is lost.

For our use case, it is best to keep the context isolated to the client's package. And use shutdown/close functions where they are required.

**Comment by omerfirmak** [comment]
I thought using context throughout the entire codebase made sense because it allows us to shutdown the entire node with a single context. It looked easier to get it right in the first try compared to per-struct Start/Shutdown functions and quit channels. I am down to anything that would get PRs merged tho :)

**Review by omerfirmak — APPROVED** [review]

**Review by IronGauntlets — APPROVED** [review]
Approving with the assumption that `ctx` will be removed in #556

**Pull request merged**